semantic_router/utils/function_call.py

```diff
@@ -6,38 +6,106 @@
 from semantic_router.llms import BaseLLM
 from semantic_router.schema import Message, RouteChoice
 from semantic_router.utils.logger import logger
+import re
 
 
-def get_schema(item: Union[BaseModel, Callable]) -> Dict[str, Any]:
-    if isinstance(item, BaseModel):
-        signature_parts = []
-        for field_name, field_model in item.__annotations__.items():
-            field_info = item.__fields__[field_name]
-            default_value = field_info.default
+def get_schemas(items: List[Union[BaseModel, Callable]]) -> List[Dict[str, Any]]:
+    schemas = []
+    for item in items:
+        if isinstance(item, BaseModel):
+            signature_parts = []
+            for field_name, field_model in item.__annotations__.items():
+                field_info = item.__fields__[field_name]
+                default_value = field_info.default
 
-            if default_value:
-                default_repr = repr(default_value)
-                signature_part = (
-                    f"{field_name}: {field_model.__name__} = {default_repr}"
-                )
-            else:
-                signature_part = f"{field_name}: {field_model.__name__}"
+                if default_value:
+                    default_repr = repr(default_value)
+                    signature_part = (
+                        f"{field_name}: {field_model.__name__} = {default_repr}"
+                    )
+                else:
+                    signature_part = f"{field_name}: {field_model.__name__}"
 
-            signature_parts.append(signature_part)
-        signature = f"({', '.join(signature_parts)}) -> str"
-        schema = {
-            "name": item.__class__.__name__,
-            "description": item.__doc__,
-            "signature": signature,
-        }
+                signature_parts.append(signature_part)
+            signature = f"({', '.join(signature_parts)}) -> str"
+            schema = {
+                "name": item.__class__.__name__,
+                "description": item.__doc__,
+                "signature": signature,
+            }
+        else:
+            schema = {
+                "name": item.__name__,
+                "description": str(inspect.getdoc(item)),
+                "signature": str(inspect.signature(item)),
+                "output": str(inspect.signature(item).return_annotation),
+            }
+        schemas.append(schema)
+    return schemas
+
+
+def convert_param_type_to_json_type(param_type: str) -> str:
+    if param_type == "int":
+        return "number"
+    if param_type == "float":
+        return "number"
+    if param_type == "str":
+        return "string"
+    if param_type == "bool":
+        return "boolean"
+    if param_type == "NoneType":
+        return "null"
+    if param_type == "list":
+        return "array"
     else:
+        return "object"
+
+
+def get_schemas_openai(items: List[Callable]) -> List[Dict[str, Any]]:
```

Comment:
We should have a single `get_schema` function, otherwise we could end up with many of these for every LLM

Instead we use the standard `get_schema` to produce the standard format, then inside the `OpenAILLM` it can convert that standard format into it's own required format — all of the other OpenAI specific functions here should also be moved there

Reply:
Makes sense to me, will do.

Reply:
thanks